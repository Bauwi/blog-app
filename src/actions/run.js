import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

// manage CRUD of run
export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

// set Run posts. This action is dispatched in the header component
// as it needs to be available for the run dropdown

export const runsHasErrored = bool => ({
  type: 'RUN_HAS_ERRORED',
  hasErrored: bool
});

export const runIsLoading = bool => ({
  type: 'RUN_IS_LOADING',
  isLoading: bool
});

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = resume => (dispatch, getState) => {
  dispatch(runIsLoading(true));
  const { uid } = getState().auth;
  const run = [];
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      return run;
    })
    .then(run => dispatch(setRunPosts(run)))
    .then(() => {
      const firstUnreadIndex = run.findIndex(post => post.state === 'unread');
      if (firstUnreadIndex === -1) {
        return dispatch(setCurrentPostRun(run[0].content.id));
      }
      return dispatch(setCurrentPostRun(run[firstUnreadIndex].content.id));
    })
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

// change the status of the past from read to unread
export const updateRunPostToAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_ALREADY_READ',
  id
});

export const startUpdateRunPostToAlreadyRead = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .update({ state: 'read' })
    .then(() => {
      dispatch(updateRunPostToAlreadyRead(id));
    });
};

// this one is not used anymore. But it could be used to mark some popst as unread yet
export const updateRunPostToNotAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ',
  id
});

// Set which post should actually be current. Run navigation depends on this.
// instead of routing, component is updated with currentPost data.
// easier way to deal with "Resume" to provide the user the first unread post in its run.
export const setCurrentPostRun = id => ({
  type: 'SET_CURRENT_POST_RUN',
  id
});

export const startSetCurrentPostRun = id => (dispatch) => {
  dispatch(runIsLoading(true));
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, 10);
  })
    .then(() => dispatch(setCurrentPostRun(id)))
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

export const resetRun = () => ({
  type: 'RESET_RUN'
});

// remove all posts in tun
export const startResetRun = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        db
          .collection('runs')
          .doc(uid)
          .collection('UserRuns')
          .doc(doc.ref.id)
          .delete();
      });
    })
    .then(() => dispatch(resetRun()));
};

// Remove already read posts in run
export const cleanRun = () => ({
  type: 'CLEAN_RUN'
});

export const startCleanRun = alreadyRead => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapShot) => {
      querySnapShot.forEach((doc) => {
        if (alreadyRead.includes(doc.ref.id)) {
          db
            .collection('runs')
            .doc(uid)
            .collection('UserRuns')
            .doc(doc.ref.id)
            .delete();
        }
      });
    })
    .then(dispatch(cleanRun()));
};
