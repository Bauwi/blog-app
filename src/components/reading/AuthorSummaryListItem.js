import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

import KeywordsList from '../KeywordsList';

export default class AuthorSummaryListItem extends Component {
  render() {
    const {
      id, title, createdAt, author, keywords, authorId, miniCover, stars
    } =
      this.props.post.content || this.props.post;
    return (
      <div>
        <Link to={`/${authorId}/read/${id}`} className="no-decoration">
          <img src={miniCover} alt="post cover" />
          <div className="masonry-layer">
            <header className="masonry-layer__header">
              <div>
                <i className="fa fa-star" /> {stars}
              </div>
              <p>{moment(createdAt).format('MMM Do, YYYY')}</p>
            </header>
            <section className="masonry-layer__section">
              <h3 className="masonry-layer__title">{title}</h3>
            </section>

            <footer className="keywords-list__list-item-container">
              <p className="masonry-layer__author">{author}</p>
              <KeywordsList keywords={keywords} />
            </footer>
          </div>
        </Link>
      </div>
    );
  }
}
