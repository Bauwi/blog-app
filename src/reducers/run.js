const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
