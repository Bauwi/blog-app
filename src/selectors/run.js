const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
