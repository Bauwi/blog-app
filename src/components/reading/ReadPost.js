import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import ReactQuill from 'react-quill';
import moment from 'moment';

import UserCard from '../UserCard';
import LoadingPage from '../LoadingPage';
import KeywordsList from '../KeywordsList';

import { startSetOnePost } from '../../actions/readings';
import { startUpPostStar } from '../../actions/posts';

export class ReadPost extends Component {
  static defaultProps = {
    post: {
      title: '',
      cover: '',
      body: [],
      keywords: ''
    }
  };
  componentDidMount() {
    this.props.startSetOnePost(this.props.postId);
  }

  onAddStar = () => {
    const { postId } = this.props;
    const { authorId } = this.props.post;
    const { stars } = this.props.author;
    this.props.startUpPostStar(postId, authorId, stars);
  };

  render() {
    if (this.props.isLoading) {
      return <LoadingPage />;
    }
    const { title, cover, body, keywords, stars, readingTime, createdAt } = this.props.post;
    const category = keywords.split(',')[0].trim();
    return (
      <div>
        <div>
          <header className="page-header page-header--read">
            <div className="content-container">
              <UserCard post={this.props.post} author={this.props.author} />
              <div className={`read-header__datereadingtime border-top-item-${category}`}>
                <p>{moment(createdAt).format('MMM Do, YYYY')}</p>
                <p>
                  <i className="fa fa-clock-o" />{' '}
                  {`${readingTime} ${readingTime >= 2 ? 'minutes' : 'minute'} read`}
                </p>
                <div className="read-header__stars">
                  <button className="button button--star" onClick={this.onAddStar}>
                    {stars}
                    <i className="fa fa-star-o" />
                  </button>
                  <p>{this.props.postStars}</p>
                </div>
              </div>
            </div>
          </header>
          <h2 className="read-header__title">{title}</h2>

          <img className="image image--cover" src={cover} />
          <div className="content-container read-only">
            <ReactQuill
              theme="snow"
              value={{ ops: [...body] }}
              readOnly
              modules={ReadPost.modules}
              bounds=".app"
              placeholder="Add a post to your blog"
            />
          </div>

          <footer className={`keywords-list__read-post-container border-top-item-${category}`}>
            <KeywordsList keywords={keywords} />
          </footer>
        </div>
      </div>
    );
  }
}

ReadPost.propTypes = {
  post: PropTypes.object,
  author: PropTypes.object,
  postId: PropTypes.string.isRequired,
  isLoading: PropTypes.bool.isRequired,
  startSetOnePost: PropTypes.func.isRequired,
  startUpPostStar: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  startSetOnePost: id => dispatch(startSetOnePost(id)),
  startUpPostStar: (id, authorId, stars) => dispatch(startUpPostStar(id, authorId, stars))
});

const mapStateToProps = (state, props) => ({
  post: state.readings.current,
  postId: props.match.params.id,
  author: state.users.author,
  isLoading: state.readings.isLoading
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ReadPost));

ReadPost.modules = {
  toolbar: []
};
