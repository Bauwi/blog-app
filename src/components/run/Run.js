import React, { Component } from 'react';
import { connect } from 'react-redux';

import InlineList from '../InlineList';
import LoadingPage from '../LoadingPage';
import { startSetRunPosts } from '../../actions/run';

export class Run extends Component {
  state = {
    loading: true
  };
  componentWillMount() {
    if (this.props.run.length === 0) {
      this.props.startSetRunPosts().then(() => {
        this.setState(() => ({ loading: false }));
      });
    } else {
      this.setState(() => ({ loading: false }));
    }
  }

  render() {
    if (this.state.loading) {
      return <LoadingPage />;
    }
    return (
      <div>
        <p>Run</p>
        <InlineList posts={this.props.run} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSetRunPosts: () => dispatch(startSetRunPosts())
});

const mapStateToProps = state => ({
  run: state.run
});

export default connect(mapStateToProps, mapDispatchToProps)(Run);
