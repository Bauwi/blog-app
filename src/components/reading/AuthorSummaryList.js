import React, { Component } from 'react';

import AuthorSummaryListItem from './AuthorSummaryListItem';

export default class AuthorSummaryList extends Component {
  renderList() {
    return this.props.posts.map((post) => {
      const key = post.content ? post.content.id : post.id;

      return <AuthorSummaryListItem key={key} post={post} />;
    });
  }

  render() {
    return (
      <div className="content-container">
        <div className="masonry">{this.renderList()}</div>;
      </div>
    );
  }
}
