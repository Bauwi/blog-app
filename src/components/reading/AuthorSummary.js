import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import UserCard from '../UserCard';
import CompleteFiltersBar from '../filters/CompleteFiltersBar';
import AuthorSummaryList from './AuthorSummaryList';

import selectPosts from '../../selectors/posts';
import LoadingPage from '../LoadingPage';

import { startSetSpecificUserPosts } from '../../actions/readings';

export class AuthorSummary extends Component {
  componentDidMount() {
    this.props.startSetSpecificUserPosts(this.props.userId);
  }

  render() {
    if (this.props.isLoading) {
      return <LoadingPage />;
    }
    return (
      <div>
        <div className="page-container">
          <div className="content-container">
            <UserCard author={this.props.author} userId={this.props.userId} />
            <CompleteFiltersBar />
          </div>
          <AuthorSummaryList posts={this.props.posts} />
        </div>
      </div>
    );
  }
}

AuthorSummary.propTypes = {
  posts: PropTypes.arrayOf(PropTypes.object).isRequired,
  author: PropTypes.object,
  userId: PropTypes.string.isRequired,
  isLoading: PropTypes.bool,
  startSetSpecificUserPosts: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  startSetSpecificUserPosts: id => dispatch(startSetSpecificUserPosts(id))
});

const mapStateToProps = (state, props) => ({
  posts: selectPosts(state.readings.currentUserPosts, state.filters),
  author: state.users.author,
  isLoading: state.readings.isLoading,
  userId: props.location.pathname.slice(1, 29)
});

export default connect(mapStateToProps, mapDispatchToProps)(AuthorSummary);
