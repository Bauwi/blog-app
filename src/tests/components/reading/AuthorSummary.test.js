import React from 'react';
import { shallow } from 'enzyme';
import { AuthorSummary } from '../../../components/reading/AuthorSummary';
import run from '../../fixtures/run';

const posts = run.posts.map(post => post.content);

test('should render AuthorSummary component properly', () => {
  const wrapper = shallow(<AuthorSummary
    isLoading={false}
    userId="1"
    posts={posts}
    startSetSpecificUserPosts={() => {}}
  />);
  expect(wrapper).toMatchSnapshot();
});

test('should render AuthorSummary component properly when loading', () => {
  const wrapper = shallow(<AuthorSummary isLoading userId="1" posts={posts} startSetSpecificUserPosts={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

test('should call startSetSpecificUserPosts on mount', () => {
  const startSetSpecificUserPosts = jest.fn();
  const wrapper = shallow(<AuthorSummary
    isLoading={false}
    userId="1"
    posts={posts}
    startSetSpecificUserPosts={startSetSpecificUserPosts}
  />);
  expect(startSetSpecificUserPosts).toHaveBeenLastCalledWith('1');
});
