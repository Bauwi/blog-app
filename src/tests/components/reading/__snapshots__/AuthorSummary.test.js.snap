// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AuthorSummary component properly 1`] = `
<div>
  <div
    className="page-container"
  >
    <div
      className="content-container"
    >
      <UserCard
        author={
          Object {
            "avatar": "https://www.timeshighereducation.com/sites/default/files/byline_photos/default-avatar.png",
            "description": "I had three chairs in my house; one for solitude, two for friendship, three for society.",
            "stars": 0,
            "username": "Anonymous",
          }
        }
        userId="1"
      />
      <PostListFilters />
    </div>
    <AuthorSummaryList
      posts={
        Array [
          Object {
            "author": "author1",
            "authorId": "author1Id",
            "body": Array [
              Object {
                "attributes": "h1",
                "text": "This is the first Post in run",
              },
            ],
            "cover": "https://www.w3schools.com/w3css/img_fjords.jpg",
            "createdAt": 0,
            "id": "1",
            "keywords": "red, green, blue",
            "miniCover": "https://www.w3schools.com/w3css/img_fjords.jpg",
            "readingTime": 7,
            "stars": 5,
            "title": "Post 1 in Run",
          },
          Object {
            "author": "author2",
            "authorId": "author2Id",
            "body": Array [
              Object {
                "attributes": "h2",
                "text": "This is the second Post in run",
              },
            ],
            "cover": "https://www.w3schools.com/w3css/img_fjords.jpg",
            "createdAt": -1500,
            "id": "2",
            "keywords": "cat, dog, fish",
            "miniCover": "https://www.w3schools.com/w3css/img_fjords.jpg",
            "readingTime": 12,
            "stars": 25,
            "title": "Post 2 in Run",
          },
          Object {
            "author": "author3",
            "authorId": "author3Id",
            "body": Array [
              Object {
                "attributes": "h3",
                "text": "blabla",
              },
            ],
            "cover": "https://www.w3schools.com/w3css/img_fjords.jpg",
            "createdAt": 1000000,
            "id": "3",
            "keywords": "car, pool, cake",
            "miniCover": "https://www.w3schools.com/w3css/img_fjords.jpg",
            "readingTime": 2,
            "stars": 2,
            "title": "Post 3 in Run",
          },
        ]
      }
    />
  </div>
</div>
`;

exports[`should render AuthorSummary component properly when loading 1`] = `<Component />`;
